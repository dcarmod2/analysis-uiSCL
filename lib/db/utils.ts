import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollection(obj) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollection)
  if (typeof obj === 'object') return mapValues(obj, serializeCollection)
  return obj
}
