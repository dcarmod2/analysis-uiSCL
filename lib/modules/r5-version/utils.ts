import {MINIMUM_R5_VERSION, VERSION_PARSE_REGEX} from './constants'

/**
 * Convert an r5 version to a number for comparing
 */
export function versionToNumber(version: string): number {
  const versionMatch = (version || '').match(VERSION_PARSE_REGEX)
  if (versionMatch) {
    const [, major, minor, patch, commit] = versionMatch

    // NB Assumes no major, minor, patch, or commit version exceeds 100
    return (
      parseInt(major) * 10000 +
      parseInt(minor) * 100 +
      parseInt(patch) +
      parseInt(commit || '0') / 100
    )
  }

  return 0
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = versionToNumber(version) % 1 !== 0
  return isTestVersion || workerVersionInRange(version, min, max)
}
