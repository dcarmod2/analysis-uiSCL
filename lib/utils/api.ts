// Query values are typed as string OR string[]
export function getQueryAsString(p: string | string[]): string {
  return Array.isArray(p) ? p[0] : p
}

// Get a query as an object
export function getQueryAsObject(
  p: string | string[]
): Record<string, unknown> {
  try {
    return JSON.parse(getQueryAsString(p))
  } catch (e) {
    return {}
  }
}

// Convert an error to POJO for sending over the wire
export function errorToPOJO(e: Error): Record<string, unknown> {
  const obj = {}
  Object.getOwnPropertyNames(e).forEach((k) => {
    obj[k] = e[k]
  })
  return obj
}
