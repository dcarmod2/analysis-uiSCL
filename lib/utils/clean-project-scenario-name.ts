//
import message from 'lib/message'

import {UNDEFINED_PROJECT_NAME} from '../constants'

export default function cleanProjectScenarioName(project, scenarioIndex) {
  if (!project) {
    return UNDEFINED_PROJECT_NAME
  }
  const scenarioName =
    project.variants[scenarioIndex] || message('variant.baseline')
  const description = `${project.name}-${scenarioName}`
  return description.replace(/[^a-zA-Z0-9]/, '-')
}
