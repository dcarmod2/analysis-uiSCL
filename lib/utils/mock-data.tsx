import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'
import React from 'react'
import {Provider} from 'react-redux'
import configureStore from 'redux-mock-store'
import thunkMiddleware from 'redux-thunk'
import {TextEncoder} from 'util'
import {constructor as XorShift} from 'xorshift'

import {responseToSurface} from '../actions/analysis'
import {ADD_TRIP_PATTERN, TRAVEL_TIME_PERCENTILES} from '../constants'

import multi from './multi'

// Create a random number generator that is seeded
const seededRandom = new XorShift([1, 2, 3, 4])

export const makeMockStore = configureStore([multi, thunkMiddleware])

export const mockWithProvider = (children, mockData = mockStores.init) => {
  const store = makeMockStore(mockData)
  const wrapper = mount(<Provider store={store}>{children}</Provider>)

  return {
    snapshot: () => mountToJson(wrapper),
    store,
    wrapper
  }
}

export const mockModel = {
  accessGroup: 'accessGroupName',
  createdAt: '2017-11-08T12:30:50.133Z',
  createdBy: 'person@gmail.com',
  nonce: '12345',
  updatedAt: '2017-11-08T12:31:19.602Z',
  updatedBy: 'person@gmail.com'
}

export const mockSegmentGeom: GeoJSON.LineString = {
  type: 'LineString',
  coordinates: [
    [-122.0246, 36.9707],
    [-122.0279, 37.049],
    [-121.9799, 37.2299],
    [-121.9445, 37.324],
    [-121.936, 37.353],
    [-121.924, 37.365]
  ]
}

export const mockSegment: CL.ModificationSegment = {
  fromStopId: '1',
  toStopId: null,
  geometry: mockSegmentGeom,
  spacing: 3000,
  stopAtEnd: false,
  stopAtStart: false
}

export const mockStop1 = {
  autoCreated: false,
  bearing: 0,
  center: [36.9707, -122.0246],
  distanceFromStart: 0,
  index: 0,
  lat: 36.9707,
  lon: -122.0246,
  stopId: '1'
}

export const mockGtfsStop1 = {
  stop_id: '1',
  stop_lat: 36.9707,
  stop_lon: -122.0246,
  stop_name: 'Santa Cruz'
}

const mockGtfsStop2 = {
  stop_id: '2',
  stop_lat: 37.049,
  stop_lon: -122.0279,
  stop_name: 'Scotts Valley'
}

export const mockGtfsStops = [mockGtfsStop1, mockGtfsStop2]

export const mockPattern = {
  geometry: mockSegmentGeom,
  name: 'Mock Pattern',
  pattern_id: 'mock-pattern-id',
  stops: mockGtfsStops,
  trips: [
    {
      duration: 1234,
      end_time: 13579,
      start_time: 12345,
      trip_headsign: 'To Downtown',
      trip_id: 'abcd',
      trip_short_name: 'The Express'
    }
  ]
}

export const mockRoute = {
  label: 'mock-route-label',
  patterns: [mockPattern],
  route_id: 'route1'
}

export const mockFeed = {
  checksum: 'MOCK',
  id: '1',
  routes: [mockRoute],
  stopsById: {
    1: mockGtfsStop1,
    2: mockGtfsStop2
  },
  stops: [mockGtfsStop1, mockGtfsStop2]
}

export const mockMapState = {
  allowExtend: true,
  extendFromEnd: true,
  followRoad: true,
  state: 'add-trip-pattern',
  modificationId: '1234'
}

export const mockTimetable = {
  _id: 'timetable-id',
  name: 'Test timetable',
  segmentSpeeds: [],
  dwellTime: 10,
  dwellTimes: [12],
  monday: true,
  tuesday: true,
  wednesday: true,
  thursday: true,
  friday: true,
  saturday: false,
  sunday: false,
  headwaySecs: 900, // 15 minutes
  startTime: 28800, // 8am
  endTime: 57600, // 4pm
  patternTrips: ['abcd'],
  sourceTrip: 'abcd',
  phaseFromTimetable: '',
  phaseFromStop: '',
  phaseAtStop: '',
  phaseSeconds: 300,
  exactTimes: false,
  modificationId: '1234',
  modificationName: 'Test Modification'
}

export const mockOpportunityDataset = {
  sourceName: 'lodes-data-2014',
  name: 'Workers total',
  _id: 'Workers_total'
}

const regionId = '1'

export const mockBundle = {
  ...mockModel,
  errorCode: '',
  name: 'Mock Bundle',
  _id: '1',
  centerLat: 38.90124,
  centerLon: -77.015615,
  feeds: [],
  regionId,
  serviceStart: '2018-09-27',
  serviceEnd: '2018-09-28',
  status: 'DONE',
  totalFeeds: 0
}

export const mockRegion = {
  ...mockModel,
  bounds: {
    north: 39,
    east: -76,
    south: 38,
    west: -77
  },
  description: 'Region description',
  _id: regionId,
  name: 'Mock Region',
  opportunityDatasets: [mockOpportunityDataset],
  statusCode: 'DONE'
}

export const mockProject = {
  ...mockModel,
  bundleId: '1',
  _id: '1',
  name: 'Mock Project',
  regionId: mockRegion._id,
  variants: ['Default']
}

export const mockModification = {
  ...mockModel,
  bidirectional: false,
  description: 'Mock description',
  entries: [],
  feed: mockFeed.id,
  _id: '1234',
  name: 'Test Modification',
  routes: [mockRoute.route_id],
  projectId: mockProject._id,
  segments: [mockSegment],
  timetables: [mockTimetable],
  trips: ['abcd'],
  type: ADD_TRIP_PATTERN,
  variants: [true]
}

export const mockProfileRequest = {
  accessModes: 'WALK',
  bikeSafe: 1,
  bikeSlope: 1,
  bikeSpeed: 4.166666666666667,
  bikeTime: 1,
  bounds: mockRegion.bounds,
  bundleId: mockBundle._id,
  date: '2016-01-16',
  defaultExpanded: true,
  directModes: 'WALK',
  egressModes: 'WALK',
  fromLat: 12,
  fromLon: 34,
  fromTime: 25200,
  maxBikeTime: 20,
  maxRides: 4,
  maxTripDurationMinutes: 60,
  maxWalkTime: 20,
  monteCarloDraws: 200,
  name: 'Profile Request Test',
  projectId: mockProject._id,
  reachabilityThreshold: 0,
  regionId: mockRegion._id,
  toTime: 32400,
  transitModes: 'BUS,TRAM,RAIL,SUBWAY',
  travelTimePercentile: 50,
  variantIndex: 0,
  walkSpeed: 1.3888888888888888,
  workerVersion: 'v4.0.0'
}

export const mockRegionalAnalyses = [
  {
    ...mockModel,
    zoom: 9,
    bounds: {},
    complete: true,
    deleted: false,
    destinationPointSetIds: ['Jobs_total'],
    status: 'DONE',
    workerVersion: 'v3.2.0',
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 1',
    grid: 'Jobs_total',
    projectId: '1',
    variant: 0,
    bundleId: '1',
    cutoffMinutes: 60,
    cutoffsMinutes: [60],
    _id: 'abcd',
    request: mockProfileRequest,
    // percentile -1 indicates this analysis was done using Andrew Owen-style average instantaneous accessibility
    travelTimePercentile: -1,
    travelTimePercentiles: [50]
  },
  {
    ...mockModel,
    zoom: 9,
    bounds: {},
    complete: true,
    deleted: false,
    destinationPointSetIds: ['Jobs_total'],
    status: 'DONE',
    workerVersion: 'v3.2.0',
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 2',
    grid: 'Jobs_total',
    projectId: '1',
    variant: 1,
    bundleId: '1',
    cutoffMinutes: 60,
    cutoffsMinutes: [60],
    _id: 'efgh',
    request: mockProfileRequest,
    travelTimePercentile: 50,
    travelTimePercentiles: [50]
  }
]

const GRID_SIZE = 300
const GRID_MAX = 2000
export const mockGrids = mockRegionalAnalyses.map((ra) => ({
  analysisId: ra._id,
  percentile: 50,
  cutoff: 60,
  zoom: 9,
  width: GRID_SIZE,
  height: GRID_SIZE,
  north: 20024,
  west: 20015,
  data: Array(GRID_SIZE * GRID_SIZE)
    .fill(0)
    .map(() => Math.floor(seededRandom.random() * GRID_MAX)),
  min: 0,
  max: GRID_MAX,
  contains: () => true
}))

export const mockGrid = mockGrids[0]

export const mockStores = {
  init: {
    analysis: {
      copyRequestSettings: true,
      isochroneFetchStatus: false,
      regional: {
        bounds: mockRegion.bounds
      },
      requestsSettings: [mockProfileRequest],
      resultsSettings: [],
      maxTripDurationMinutes: 60,
      travelTimePercentile: 50
    },
    network: {
      error: null,
      outstandingRequests: 0
    },
    opportunityDatasets: {
      active: mockOpportunityDataset._id,
      datasets: mockRegion.opportunityDatasets,
      uploadStatuses: []
    },
    queryString: {
      analysisId: mockRegionalAnalyses[0]._id,
      cutoff: 60,
      percentile: 50,
      modificationId: mockModification._id,
      projectId: mockProject._id,
      regionId: mockRegion._id
    },
    region: {
      aggregationAreas: [],
      bundles: [mockBundle],
      regions: [mockRegion]
    },
    regionalAnalyses: {
      analyses: mockRegionalAnalyses,
      grids: mockGrids
    },
    project: {
      activeModificationId: mockModification._id,
      currentProject: mockProject,
      feeds: [mockFeed],
      modifications: [mockModification],
      projects: [mockProject]
    },
    user: {
      idToken: null
    }
  }
}

export const mockScenarioApplicationError = {
  modificationId: 'Add Trip Pattern',
  title: "Error while applying the modification entitled 'Add Trip Pattern'.",
  messages: ['Headway is not greater than zero.']
}

export const mockSurfaceParameters = {
  width: 5,
  height: 5,
  depth: 5,
  zoom: 9,
  north: 40000,
  west: 40000
}

/**
 * This is a function because responseToSurface is destructive and can lead to
 * test leakage if we don't have a brand-new surface each time.
 */
export const createMockSurfaceData = () => {
  const {width, height, depth, zoom, north, west} = mockSurfaceParameters

  const metadata = {
    scenarioApplicationWarnings: [mockScenarioApplicationError]
  }

  const metadataEncoded = new TextEncoder().encode(JSON.stringify(metadata))
  const bufSize =
    9 * 4 + // header
    width * height * depth * 4 + // data
    metadataEncoded.length // metadata

  const buffer = new ArrayBuffer(bufSize)
  const surfaceData = new Int32Array(buffer, 0, 9 + width * height * depth)

  // text header ("ACCESSGR")
  surfaceData[0] = 1162036033 // ACCE
  surfaceData[1] = 1380406099 // SSGR

  // version
  surfaceData[2] = 0

  surfaceData[3] = zoom
  surfaceData[4] = west
  surfaceData[5] = north
  surfaceData[6] = width
  surfaceData[7] = height
  surfaceData[8] = depth

  // phew, header is done
  for (let z = 0, position = 9; z < depth; z++) {
    let prev = 0
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++, position++) {
        const pixelVal = getMockSurfaceValue(x, y, z)
        surfaceData[position] = pixelVal - prev // values are delta-coded within each depth
        prev = pixelVal
      }
    }
  }

  // copy metadata
  const metadataBuffer = new Uint8Array(
    buffer,
    buffer.byteLength - metadataEncoded.byteLength,
    metadataEncoded.length
  )
  for (let i = 0; i < metadataEncoded.length; i++) {
    metadataBuffer[i] = metadataEncoded[i]
  }

  return buffer
}

export const mockSurface = responseToSurface(createMockSurfaceData())

/**
 * Return the value stored in sample z of pixel x, y of the mock surface
 */
export function getMockSurfaceValue(x, y, z) {
  // This used to simply return z, 2 * z and 3 * z when we had an array of
  // [5, 10, 15 . . . 95]th percentiles. However, when we switched to
  // requesting specific percentiles that stopped working. This translates the
  // position in the access grid into a percentile, then subtracts five and
  // divides by five to get the position in the old array, so that tests
  // written assuming the old structure continue to pass.
  if (x < 2 && y < 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y >= 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y < 2) return (2 * (TRAVEL_TIME_PERCENTILES[z] - 5)) / 5
  if (x < 2 && y >= 2) return (3 * (TRAVEL_TIME_PERCENTILES[z] - 5)) / 5
  return 0
}

const opportunityDatasetGridData = new Int32Array(
  mockSurfaceParameters.width * mockSurfaceParameters.height
)
// top half of grid is 1, bottom half is 2
for (let i = 0; i < 3 * mockSurfaceParameters.width; i++) {
  opportunityDatasetGridData[i] = 1
}
for (
  let i = 3 * mockSurfaceParameters.width;
  i < opportunityDatasetGridData.length;
  i++
) {
  opportunityDatasetGridData[i] = 2
}

export const mockDestinationGrid = {
  ...mockSurfaceParameters,
  west: mockSurfaceParameters.west - 1, // shift slightly so that we are testing misaligned grids
  north: mockSurfaceParameters.north - 1,
  data: opportunityDatasetGridData,
  getValue(x, y) {
    return opportunityDatasetGridData[y * mockSurfaceParameters.width + x]
  }
}

export const releaseVersions = ['v3.0.0', 'v2.2.0', 'v2.0.0']
export const allVersions = [
  'v3.0.0',
  'v2.2.0-2-g789df3e',
  'v2.2.0',
  'v2.0.0-19-g589f32f',
  'v2.0.0'
]
