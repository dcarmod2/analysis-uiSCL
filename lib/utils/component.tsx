import {ThemeProvider} from '@chakra-ui/core'
import {expect} from '@jest/globals'
import enzyme from 'enzyme'
import {Router} from 'next/router'
import {RouterContext} from 'next/dist/next-server/lib/router-context'
import React from 'react'
import {Map} from 'react-leaflet'
import {Provider} from 'react-redux'

import {makeMockStore, mockStores} from './mock-data'

// Next.js uses Math.random to generate unique ids for pages that get captured in snapshots.
global.Math.random = () => 0.5

// Set __NEXT_DATA__ (required by Router)
window.__NEXT_DATA__ = {} as any
const router = new Router('page', {}, 'as-path', {} as any)

const getName = (c) => c.displayName || c.name

export function testComponent(
  TestComponent,
  props = {},
  initialData = mockStores.init
) {
  const store = makeMockStore(initialData)
  const Wrapping = (props) => (
    <Provider store={store}>
      <ThemeProvider>
        <RouterContext.Provider value={router}>
          {props.children}
        </RouterContext.Provider>
      </ThemeProvider>
    </Provider>
  )

  const mount = () =>
    enzyme.mount(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // TODO: Dive down through the `TestComponent` automatically
  const shallow = () =>
    enzyme.shallow(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // For debugging purposes (this is not the original component!)
  const name = `WithContext(${getName(TestComponent)})`

  return {mount, name, router, shallow, store}
}

/**
 * Rename to `wrapComponent`. TODO replace all `testComponent`
 */
export const wrapComponent = testComponent

/**
 * Helper function to create quick snapshot tests for components.
 */
export function testAndSnapshot(C, p, id?: any) {
  test(`Component ${getName(C)} snapshot(mount)`, () => {
    const c = testComponent(C, p, id)
    const t = c.mount()
    expect(t).toMatchSnapshot()
    t.unmount()
  })
}

/**
 * Test and snapshot map component
 */
export function wrapMapComponent(C, p, id?: any) {
  const MapWrapper = (props) => (
    <Map>
      <C {...props} />
    </Map>
  )
  MapWrapper.displayName = `WithMap(${getName(C)})`
  return testComponent(MapWrapper, p, id)
}

function snapshotTest(wc, type) {
  test(`Component ${getName(wc)} snapshot(${type})`, () => {
    const s = wc[`${type}`]()
    expect(s).toMatchSnapshot()
    s.unmount()
  })
}

/**
 * Shallow snapshot test.
 */
export function shallowSnapshot(wc) {
  snapshotTest(wc, 'shallow')
}

/**
 * Deep snapshot test.
 */
export function deepSnapshot(wc) {
  snapshotTest(wc, 'mount')
}
