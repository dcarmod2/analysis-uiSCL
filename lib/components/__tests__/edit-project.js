import {mockBundle, mockProject} from 'lib/utils/mock-data'
import {testComponent} from 'lib/utils/component'

import EditProject from '../edit-project'

test('Component > EditProject', () => {
  const props = {
    bundleName: mockBundle.name,
    project: mockProject,
    query: {}
  }

  const c = testComponent(EditProject, props)
  expect(c.mount()).toMatchSnapshot()
})
