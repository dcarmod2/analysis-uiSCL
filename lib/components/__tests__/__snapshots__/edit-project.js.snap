// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component > EditProject 1`] = `
<EditProject
  bundleName="Mock Bundle"
  project={
    Object {
      "_id": "1",
      "accessGroup": "accessGroupName",
      "bundleId": "1",
      "createdAt": "2017-11-08T12:30:50.133Z",
      "createdBy": "person@gmail.com",
      "name": "Mock Project",
      "nonce": "12345",
      "regionId": "1",
      "updatedAt": "2017-11-08T12:31:19.602Z",
      "updatedBy": "person@gmail.com",
      "variants": Array [
        "Default",
      ],
    }
  }
  query={Object {}}
>
  <InnerDock>
    <SimpleBar
      className="InnerDock "
      style={
        Object {
          "height": 768,
        }
      }
    >
      <div
        className="InnerDock "
        data-simplebar={true}
        style={
          Object {
            "height": 768,
          }
        }
      >
        <div
          className="simplebar-wrapper"
        >
          <div
            className="simplebar-height-auto-observer-wrapper"
          >
            <div
              className="simplebar-height-auto-observer"
            />
          </div>
          <div
            className="simplebar-mask"
          >
            <div
              className="simplebar-offset"
            >
              <div
                className="simplebar-content-wrapper"
              >
                <div
                  className="simplebar-content"
                >
                  <ForwardRef
                    p={4}
                    spacing={4}
                  >
                    <Flex
                      direction="column"
                      p={4}
                    >
                      <Box
                        display="flex"
                        flexDirection="column"
                        p={4}
                      >
                        <div
                          className="css-1n0jko5"
                        >
                          <Heading
                            key=".0"
                            mb={4}
                            size="md"
                          >
                            <Box
                              as="h2"
                              fontFamily="heading"
                              fontSize="xl"
                              fontWeight="bold"
                              lineHeight="shorter"
                              mb={4}
                            >
                              <h2
                                className="css-k9oqgk"
                              >
                                Edit project settings
                              </h2>
                            </Box>
                          </Heading>
                          <FormControl
                            isInvalid={false}
                            key=".1"
                            mb={4}
                          >
                            <Box
                              mb={4}
                              role="group"
                            >
                              <div
                                className="css-acwcvw"
                                role="group"
                              >
                                <FormLabel
                                  htmlFor="projectName"
                                >
                                  <Box
                                    as="label"
                                    display="inline-block"
                                    fontSize="md"
                                    fontWeight="medium"
                                    htmlFor="projectName"
                                    opacity="1"
                                    pb="4px"
                                    pr="12px"
                                    textAlign="left"
                                    verticalAlign="middle"
                                  >
                                    <label
                                      className="css-176s5fq"
                                      htmlFor="projectName"
                                    >
                                      Project name
                                    </label>
                                  </Box>
                                </FormLabel>
                                <Input
                                  as="input"
                                  errorBorderColor="red.500"
                                  focusBorderColor="blue.500"
                                  id="projectName"
                                  isFullWidth={true}
                                  onChange={[Function]}
                                  size="md"
                                  value="Mock Project"
                                  variant="outline"
                                >
                                  <PseudoBox
                                    _disabled={
                                      Object {
                                        "cursor": "not-allowed",
                                        "opacity": "0.4",
                                      }
                                    }
                                    _focus={
                                      Object {
                                        "borderColor": "#3182ce",
                                        "boxShadow": "0 0 0 1px #3182ce",
                                        "zIndex": 1,
                                      }
                                    }
                                    _hover={
                                      Object {
                                        "borderColor": "gray.300",
                                      }
                                    }
                                    _invalid={
                                      Object {
                                        "borderColor": "#e53e3e",
                                        "boxShadow": "0 0 0 1px #e53e3e",
                                      }
                                    }
                                    _readOnly={
                                      Object {
                                        "bg": "transparent",
                                        "boxShadow": "none !important",
                                        "userSelect": "all",
                                      }
                                    }
                                    alignItems="center"
                                    appearance="none"
                                    aria-invalid={false}
                                    as="input"
                                    bg="white"
                                    border="1px"
                                    borderColor="inherit"
                                    display="flex"
                                    fontSize="md"
                                    height={10}
                                    id="projectName"
                                    onChange={[Function]}
                                    outline="none"
                                    position="relative"
                                    px={4}
                                    rounded="md"
                                    transition="all 0.2s"
                                    value="Mock Project"
                                    width="100%"
                                  >
                                    <input
                                      aria-invalid={false}
                                      className="css-1ga4kqw"
                                      id="projectName"
                                      onChange={[Function]}
                                      value="Mock Project"
                                    />
                                  </PseudoBox>
                                </Input>
                              </div>
                            </Box>
                          </FormControl>
                          <ForwardRef
                            key=".2"
                            mb={4}
                            spacing={2}
                          >
                            <Flex
                              direction="column"
                              mb={4}
                            >
                              <Box
                                display="flex"
                                flexDirection="column"
                                mb={4}
                              >
                                <div
                                  className="css-opcg3f"
                                >
                                  <Heading
                                    key=".0"
                                    mb={2}
                                    size="sm"
                                  >
                                    <Box
                                      as="h2"
                                      fontFamily="heading"
                                      fontSize="md"
                                      fontWeight="bold"
                                      lineHeight="shorter"
                                      mb={2}
                                    >
                                      <h2
                                        className="css-zngbqj"
                                      >
                                        Bundle
                                      </h2>
                                    </Box>
                                  </Heading>
                                  <Box
                                    key=".1"
                                    mb={null}
                                  >
                                    <div
                                      className="css-0"
                                    >
                                      Bundle cannot be changed once a project is created. 
                                      <ALink
                                        bundleId="1"
                                        regionId="1"
                                        to="bundleEdit"
                                      >
                                        <InternalLink
                                          bundleId="1"
                                          regionId="1"
                                          to="bundleEdit"
                                        >
                                          <Link
                                            as="/regions/1/bundles/1"
                                            href={
                                              Object {
                                                "pathname": "/regions/[regionId]/bundles/[bundleId]",
                                                "query": Object {},
                                              }
                                            }
                                            passHref={true}
                                          >
                                            <Link
                                              _hover={
                                                Object {
                                                  "color": "blue.700",
                                                }
                                              }
                                              className=""
                                              color="blue.500"
                                              href="/regions/1/bundles/1"
                                              onClick={[Function]}
                                              onMouseEnter={[Function]}
                                            >
                                              <PseudoBox
                                                _disabled={
                                                  Object {
                                                    "cursor": "not-allowed",
                                                    "opacity": "0.4",
                                                    "textDecoration": "none",
                                                  }
                                                }
                                                _focus={
                                                  Object {
                                                    "boxShadow": "outline",
                                                  }
                                                }
                                                _hover={
                                                  Object {
                                                    "color": "blue.700",
                                                  }
                                                }
                                                as="a"
                                                className=""
                                                color="blue.500"
                                                cursor="pointer"
                                                href="/regions/1/bundles/1"
                                                onClick={[Function]}
                                                onMouseEnter={[Function]}
                                                outline="none"
                                                textDecoration="none"
                                                transition="all 0.15s ease-out"
                                              >
                                                <a
                                                  className=" css-17d37xz"
                                                  href="/regions/1/bundles/1"
                                                  onClick={[Function]}
                                                  onMouseEnter={[Function]}
                                                >
                                                  View bundle info here.
                                                </a>
                                              </PseudoBox>
                                            </Link>
                                          </Link>
                                        </InternalLink>
                                      </ALink>
                                    </div>
                                  </Box>
                                </div>
                              </Box>
                            </Flex>
                          </ForwardRef>
                          <Button
                            isDisabled={true}
                            key=".3"
                            leftIcon="edit"
                            mb={4}
                            onClick={[Function]}
                            variantColor="green"
                          >
                            <PseudoBox
                              _active={
                                Object {
                                  "bg": "green.700",
                                }
                              }
                              _disabled={
                                Object {
                                  "boxShadow": "none",
                                  "cursor": "not-allowed",
                                  "opacity": "40%",
                                }
                              }
                              _focus={
                                Object {
                                  "boxShadow": "outline",
                                }
                              }
                              _hover={
                                Object {
                                  "bg": "green.600",
                                }
                              }
                              alignItems="center"
                              appearance="none"
                              aria-disabled={true}
                              as="button"
                              bg="green.500"
                              borderRadius="md"
                              color="white"
                              disabled={true}
                              display="inline-flex"
                              fontSize="md"
                              fontWeight="semibold"
                              height={10}
                              justifyContent="center"
                              lineHeight="1.2"
                              mb={4}
                              minWidth={10}
                              onClick={[Function]}
                              outline="none"
                              position="relative"
                              px={4}
                              transition="all 250ms"
                              type="button"
                              userSelect="none"
                              verticalAlign="middle"
                              whiteSpace="nowrap"
                            >
                              <button
                                aria-disabled={true}
                                className="css-1egop4s"
                                disabled={true}
                                onClick={[Function]}
                                type="button"
                              >
                                <ButtonIcon
                                  icon="edit"
                                  ml={-1}
                                  mr={2}
                                >
                                  <Icon
                                    aria-hidden="true"
                                    color="currentColor"
                                    focusable="false"
                                    ml={-1}
                                    mr={2}
                                    name="edit"
                                  >
                                    <Styled(div)
                                      aria-hidden="true"
                                      as="svg"
                                      color="currentColor"
                                      display="inline-block"
                                      focusable="false"
                                      ml={-1}
                                      mr={2}
                                      role="presentation"
                                      size="1em"
                                      verticalAlign="middle"
                                      viewBox="0 0 24 24"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        className="css-yxiis9"
                                        focusable="false"
                                        role="presentation"
                                        viewBox="0 0 24 24"
                                      >
                                        <g
                                          fill="none"
                                          stroke="currentColor"
                                          strokeLinecap="full"
                                          strokeWidth="2"
                                        >
                                          <path
                                            d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
                                          />
                                          <path
                                            d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
                                          />
                                        </g>
                                      </svg>
                                    </Styled(div)>
                                  </Icon>
                                </ButtonIcon>
                                Save project
                              </button>
                            </PseudoBox>
                          </Button>
                          <ConfirmButton
                            description="Are you sure you would like to delete this project?"
                            key=".4"
                            leftIcon="delete"
                            mb={null}
                            onConfirm={[Function]}
                            variantColor="red"
                          >
                            <Button
                              leftIcon="delete"
                              mb={null}
                              onClick={[Function]}
                              variantColor="red"
                            >
                              <PseudoBox
                                _active={
                                  Object {
                                    "bg": "red.700",
                                  }
                                }
                                _disabled={
                                  Object {
                                    "boxShadow": "none",
                                    "cursor": "not-allowed",
                                    "opacity": "40%",
                                  }
                                }
                                _focus={
                                  Object {
                                    "boxShadow": "outline",
                                  }
                                }
                                _hover={
                                  Object {
                                    "bg": "red.600",
                                  }
                                }
                                alignItems="center"
                                appearance="none"
                                as="button"
                                bg="red.500"
                                borderRadius="md"
                                color="white"
                                display="inline-flex"
                                fontSize="md"
                                fontWeight="semibold"
                                height={10}
                                justifyContent="center"
                                lineHeight="1.2"
                                mb={null}
                                minWidth={10}
                                onClick={[Function]}
                                outline="none"
                                position="relative"
                                px={4}
                                transition="all 250ms"
                                type="button"
                                userSelect="none"
                                verticalAlign="middle"
                                whiteSpace="nowrap"
                              >
                                <button
                                  className="css-thv05u"
                                  onClick={[Function]}
                                  type="button"
                                >
                                  <ButtonIcon
                                    icon="delete"
                                    ml={-1}
                                    mr={2}
                                  >
                                    <Icon
                                      aria-hidden="true"
                                      color="currentColor"
                                      focusable="false"
                                      ml={-1}
                                      mr={2}
                                      name="delete"
                                    >
                                      <Styled(div)
                                        aria-hidden="true"
                                        as="svg"
                                        color="currentColor"
                                        display="inline-block"
                                        focusable="false"
                                        ml={-1}
                                        mr={2}
                                        role="presentation"
                                        size="1em"
                                        verticalAlign="middle"
                                        viewBox="0 0 24 24"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          className="css-yxiis9"
                                          focusable="false"
                                          role="presentation"
                                          viewBox="0 0 24 24"
                                        >
                                          <g
                                            fill="currentColor"
                                          >
                                            <path
                                              d="M19.452,7.5H4.547a.5.5,0,0,0-.5.545L5.334,22.181A2,2,0,0,0,7.326,24h9.347a2,2,0,0,0,1.992-1.819L19.95,8.045a.5.5,0,0,0-.129-.382A.5.5,0,0,0,19.452,7.5Zm-9.2,13a.75.75,0,0,1-1.5,0v-9a.75.75,0,0,1,1.5,0Zm5,0a.75.75,0,0,1-1.5,0v-9a.75.75,0,0,1,1.5,0Z"
                                            />
                                            <path
                                              d="M22,4H17.25A.25.25,0,0,1,17,3.75V2.5A2.5,2.5,0,0,0,14.5,0h-5A2.5,2.5,0,0,0,7,2.5V3.75A.25.25,0,0,1,6.75,4H2A1,1,0,0,0,2,6H22a1,1,0,0,0,0-2ZM9,3.75V2.5A.5.5,0,0,1,9.5,2h5a.5.5,0,0,1,.5.5V3.75a.25.25,0,0,1-.25.25H9.25A.25.25,0,0,1,9,3.75Z"
                                            />
                                          </g>
                                        </svg>
                                      </Styled(div)>
                                    </Icon>
                                  </ButtonIcon>
                                  Delete project
                                </button>
                              </PseudoBox>
                            </Button>
                          </ConfirmButton>
                        </div>
                      </Box>
                    </Flex>
                  </ForwardRef>
                </div>
              </div>
            </div>
          </div>
          <div
            className="simplebar-placeholder"
          />
        </div>
        <div
          className="simplebar-track simplebar-horizontal"
        >
          <div
            className="simplebar-scrollbar"
          />
        </div>
        <div
          className="simplebar-track simplebar-vertical"
        >
          <div
            className="simplebar-scrollbar"
          />
        </div>
      </div>
    </SimpleBar>
  </InnerDock>
</EditProject>
`;
